### PR TITLE
add cards from the new pack bran the builder and related engine changes

### PR DESCRIPTION
**Pull request by EvadLhorg**
changes made by @Patane97 and copied over from the playtesting repository

### CODE REVIEW

**Comment by EvadLhorg** [comment]
todo: update throneteki-json-data when https://github.com/throneteki/throneteki-json-data/pull/132 is merged

**Review by ystros — COMMENTED** [review]
Haven't looked at the actual card implementations yet, but have some initial comments on engine changes.

File: server/game/gamesteps/killcharacters.js [hunk 1]
```diff
@@ -10,7 +10,7 @@ class KillCharacters extends BaseStep {
 
     continue() {
         let cardsInPlay = this.cards.filter(card => card.location === 'play area');
-        this.game.applyGameAction('kill', cardsInPlay, killable => {
+        this.game.applyGameAction('killed', cardsInPlay, killable => {
```
[the remark this answers is not in the document]
Comment (EvadLhorg) [reply server/game/gamesteps/killcharacters.js:13]: looks like a change from the main repo hasn´t made it to the playtesting repo yet. i will revert this change

File: server/game/drawcard.js [hunk 2]
```diff
@@ -286,12 +287,26 @@ class DrawCard extends BaseCard {
...
-    addIcon(icon) {
-        this.icons.add(icon);
+    addIcon(icon, applying = true) {
+        if(this.icons.add(icon)) {
+            let params = {
+                card: this,
+                icon: icon,
+                applying: applying
+            };
+            this.game.raiseEvent('onIconAdded', params);
```
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/drawcard.js:297]: What would you recommend as the better approach to implement events to an icon instance being lost? (And even though it's not used, yet, I'd recommend we do the same for events to icons being added)

File: server/game/PropertyTypes/ReferenceCountedSetProperty.js [hunk 3]
```diff
@@ -6,13 +6,17 @@ class ReferenceCountedSetProperty {
     add(value) {
         let lowerCaseValue = value.toLowerCase();
         let currentCount = this.referenceCounts.get(lowerCaseValue) || 0;
-        this.referenceCounts.set(lowerCaseValue, currentCount + 1);
+        let newCount = currentCount + 1;
+        this.referenceCounts.set(lowerCaseValue, newCount);
+        return currentCount <= 0 && newCount > 0;
```
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/PropertyTypes/ReferenceCountedSetProperty.js:11]: This is outdated and can be removed.

For reference, it was when the DT was ruling reactions to icon adding or removal as when the icon itself was actually added or removed, rather than just an instance being added or removed. This code was tracking when a reference would actually be added after an instance was added by comparing it's count before and after the adjustmemt.

File: server/game/effects.js [hunk 4]
```diff
@@ -318,13 +318,23 @@ const Effects = {
             }
         };
     },
+    addInsightLimit: function(value) {
```
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/effects.js:321]: This was used for an earlier version of Sunspear & I must've simply missed this addition. I will remove it from this PR :)

File: server/game/effects.js [hunk 5]
```diff
@@ -589,6 +603,23 @@ const Effects = {
             }
         };
     },
+    moveToTopOfDeckIfStillInPlay: function(allowSave = true) {
```
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/effects.js:606]: Will remove this also; likely implemented for an earlier version of a card.

File: server/game/costs/MovePowerToOppFactionCost.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+        this.opponent = opponent;
+        this.amount = amount;
+    }
+
+    getAmountValue(context) {
+        return (this.amount === 'X' ? context.xValue : this.amount) || 1;
+    }
+
+    isEligible(card, context) {
+        return ['faction', 'play area'].includes(card.location) && card.getPower() >= this.getAmountValue(context);
```
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/costs/MovePowerToOppFactionCost.js:13]: I'd recommend we also remove this; left-over from a previous version of Storms End I believe

File: server/game/cards/22-BtB/TheMadKingsCommand.js [hunk 7]
```diff
@@ -0,0 +1,142 @@
+const PlotCard = require('../../plotcard.js');
+const GameActions = require('../../GameActions');
+const Messages = require('../../Messages');
+
+class TheMadKingsCommand extends PlotCard {
```
Comment (ystros) [inline server/game/cards/22-BtB/TheMadKingsCommand.js:5]: Come back to this one

File: server/game/cards/22-BtB/Gendry.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+                cardCondition: card => card.location === 'play area' && card.getType() === 'character' && !card.isLoyal()
+                                                        && card.attachments.some(attachment => attachment.hasTrait('Weapon'))
+            },
+            handler: context => {
+                let keywords = this.getKeywords();
+                this.untilEndOfPhase(ability => ({
+                    match: context.target,
+                    effect: [
+                        ability.effects.modifyStrength(2),
+                        flatten(keywords.map(keyword => ability.effects.addKeyword(keyword)))
```
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/cards/22-BtB/Gendry.js:25]: I used Vargo Hoat as a reference here, and can confirm that it definitely does work as intended on PT server (chosen character gets +2 STR & all of Gendry's keywords). I do understand your logic though, but not sure if it's worth an adjustment as its currently working fine, and is previously used on Vargo. :)

File: server/game/cards/22-BtB/FirstBuilder.js [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            ]
+        });
+        
+        this.reaction({
+            when: {
+                afterChallenge: event => event.challenge.winner === this.controller && event.challenge.challengeType === 'power' && this.isBuilderParticipatingInChallenge(this.controller)
+            },
+            limit: ability.limit.perPhase(1),
+            handler: () => {
+                this.game.promptForDeckSearch(this.controller, {
```
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/cards/22-BtB/FirstBuilder.js:19]: I've actually managed to do this (tested it also to confirm it is working as intended). Pushing through with latest changes :)

File: server/game/cards/22-BtB/Pyke.js [hunk 10]
```diff
@@ -0,0 +1,39 @@
+const DrawCard = require('../../drawcard.js');
+
+class Pyke extends DrawCard {
+    setupCardAbilities(ability) {
+        this.persistentEffect({
+            targetLocation: 'hand',
+            match: card => card.hasTrait('Warship'),
+            effect: ability.effects.gainAmbush(-1)
```
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/cards/22-BtB/Pyke.js:8]: Not sure if this is necessary, especially when considering it might not always be reduce (eg. if there is an ability that gives a card Ambush(X), X being its cost +1, rather than -1.

File: server/game/PropertyTypes/ReferenceCountedSetProperty.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/PropertyTypes/ReferenceCountedSetProperty.js]: This has been fixed/removed in a942bb5c0b401056a091d332f3580f7c3841603a

File: server/game/costs/MovePowerToOppFactionCost.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/costs/MovePowerToOppFactionCost.js]: This has been fixed/removed in a942bb5c0b401056a091d332f3580f7c3841603a

File: server/game/effects.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/effects.js]: This has been fixed/removed in a942bb5c0b401056a091d332f3580f7c3841603a and 19efe1ec75c18cfa83d14ff0709abcbf143d0f44

File: server/game/effects.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/effects.js]: This has been fixed/removed in a942bb5c0b401056a091d332f3580f7c3841603a

File: server/game/cards/22-BtB/BrienneTheBlue.js [hunk 11]
```diff
@@ -0,0 +1,28 @@
+const DrawCard = require('../../drawcard.js');
+
+class BrienneTheBlue extends DrawCard {
+    setupCardAbilities() {
+        this.action({
+            condition: () => this.hasSingleParticipatingChar(),
```
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/cards/22-BtB/BrienneTheBlue.js]: Fixed in 19efe1ec75c18cfa83d14ff0709abcbf143d0f44

File: server/game/cards/22-BtB/CasterlyRock.js [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+            condition: () => !this.kneeled,
+            match: card => card.controller === this.controller && card.location === 'play area' && card.isFaction('lannister'),
+            effect: ability.effects.canSpendGold(spendParams => spendParams.activePlayer === this.controller)
+        });
+        this.action({
+            title: 'Select cards to gain gold',
+            cost: ability.costs.kneelSelf(),
+            phase: 'marshal',
+            handler: context => {
+                this.game.promptForSelect(context.player, {
```
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/cards/22-BtB/CasterlyRock.js]: Fixed (I believe) in 19efe1ec75c18cfa83d14ff0709abcbf143d0f44

File: server/game/cards/22-BtB/TheMadKingsCommand.js [hunk 13]
```diff
@@ -0,0 +1,142 @@
...
+            for(let selectionTarget of selectionTargets) {
+                let targetLocation = validInPlayTypes.includes(selectionTarget) ? 'play area' : selectionTarget;
+                let cardsInLocation = this.game.allCards.filter(card => card.controller === player 
+                    && card.location === targetLocation // checking location
+                    && (validInPlayTypes.includes(selectionTarget) ? card.getType() === selectionTarget : true) // checking type, if applicable
+                    && card.allowGameAction('returnCardToDeck')); // checking return to deck
+                let playerSpecificToMove = cardsInLocation.filter(card => !selectedCards.includes(card));
+                toMove = toMove.concat(playerSpecificToMove);
+            }
+        }
```
Comment (ystros) [inline server/game/cards/22-BtB/TheMadKingsCommand.js:73]: The location and type logic, while likely correct, is hard to follow. 

I'm wondering if this would be more straightforward if you gathered all of the cards of certain types / locations, then removed the chosen ones from that list. E.g. something like (but maybe not exactly)
```javascript
const potentialCards = this.game.allCards.filter(card => (
    card.allowGameAction('returnCardToDeck') && (
        ['hand', 'shadows'].includes(card.location) || 
        card.location === 'play area' && ['character', 'location', 'attachment'].includes(card.getType())
    )
));
const selectedCards = selections.reduce((selectedCards, selection) => {
    return selectedCards.concat(selection.value || []);
}, []);
const toMove = potentialCards.filter(card => !selectedCards.includes(card));
```

File: server/game/cards/22-BtB/Gendry.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (ystros) [reply server/game/cards/22-BtB/Gendry.js:25]: Vargo Hoat has the `flatten` around the entire array, as I suggested above.

https://github.com/throneteki/throneteki/blob/f8ea72d5a7625b234f2dda60e7fcc8a9883ddc68/server/game/cards/18-FH/VargoHoat.js#L42

It appears to be working because of this line:
https://github.com/throneteki/throneteki/blob/f8ea72d5a7625b234f2dda60e7fcc8a9883ddc68/server/game/game.js#L262

I'd still recommend changing it.

File: server/game/cards/22-BtB/TheMadKingsCommand.js (hunk 13 above)
Comment (Patane97) [reply server/game/cards/22-BtB/TheMadKingsCommand.js]: Solid suggestion! The above code makes more sense to me; I've implemented & tested it, and it works as intended. Code has been pushed in e0a7c45.

It was so complicated as the selections kept changing throughout development, and I was getting tired of adjusting it. 😄

File: server/game/cards/22-BtB/Gendry.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/cards/22-BtB/Gendry.js]: Fair enough! Adjusted in 25b8058ec012ab571992dc9ee71226e70b945c3e (along with message fix).

File: server/game/CardSelector.js [hunk 14]
```diff
@@ -8,7 +8,7 @@ const UpToXCardSelector = require('./CardSelectors/UpToXCardSelector');
 const defaultProperties = {
     numCards: 1,
     cardCondition: () => true,
-    cardType: ['attachment', 'character', 'event', 'location'],
+    cardType: ['attachment', 'character', 'event', 'location', 'faction'],
```
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/CardSelector.js]: I believe this was leftover from a previous iteration of Storm's End. I have removed this adjustment in 1761ebf.

File: server/game/cards/22-BtB/Highgarden.js [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+            handler: () => {
+                this.game.promptForSelect(this.controller, {
+                    activePromptTitle: 'Reveal up to 3 cards',
+                    source: this,
+                    numCards: 3,
+                    mode: 'upTo',
+                    cardCondition: card => card.controller === this.controller && card.location === 'hand' && card.isFaction('tyrell') && card.getType() === 'character',
+                    onSelect: (player, cards) => this.revealSelect(player, cards),
+                    onCancel: (player) => this.revealCancel(player)
+                });
```
Comment (ystros) [inline server/game/cards/22-BtB/Highgarden.js:21]: Would it make sense to change this to a `target` declaration with `mode: 'select'` (since it doesn't have the word choose).  This would have the benefits of:
1. The player can't trigger the ability unless they actually intend to reveal 1 or more cards (unless revealing 0 cards is intentionally allowed by the rules, in which case the current cancel message should not be a `danger` message)
2. Can get rid of the `revealCancel` method and fold the `revealSelect` into the handler.
3. Easier set up for using GameActions in the future.

File: server/game/cards/22-BtB/Highgarden.js [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+                let str = amount * 2;
+                return { text: str.toString(), method: 'addCharacterToStrMap', arg: str };
+            });
+            this.game.promptWithMenu(player, this, {
+                activePrompt: {
+                    menuTitle: `Select +STR for ${this.remainingCards[0].name}`,
+                    buttons: buttons
+                },
+                source: this
+            });
```
Comment (ystros) [inline server/game/cards/22-BtB/Highgarden.js:112]: Would love it if the overall prompt was something like clicking a character multiple times to choose them multiple times, instead of choosing characters then button prompts for how much STR. But looking at the existing select card prompt code, this isn't easily doable. Perhaps as a future improvement.

File: server/game/cards/22-BtB/Highgarden.js [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+    charactersSelect(player, cards, numRevealed) {
+        this.strCharacterMap = new Map();
+        this.remainingCards = cards;
+        this.remainingStr = numRevealed * 2;
+        
+        this.game.queueSimpleStep(() => this.calculateNextCharacter(player));
+        
+        this.game.queueSimpleStep(() => {
+            for(let [str, cards] of this.strCharacterMap) {
```
Comment (ystros) [inline server/game/cards/22-BtB/Highgarden.js:74]: This definitely works, but it's perhaps conceptually a bit strange to use a map for this. I think a simple array with plain objects (e.g. `{ str: 2, card: Arya }`) would be more straightforward. (and probably closer to how it would work if it were a real prompt class)

File: server/game/drawcard.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (Patane97) [reply server/game/drawcard.js]: These events have been re-implemented as `GameActions.gainIcon` and `GameActions.loseIcon` in 2ff55ab.

File: server/game/cards/22-BtB/Highgarden.js (hunk 16 above)
Comment (Patane97) [reply server/game/cards/22-BtB/Highgarden.js]: I do agree here; maybe a 'repeatable' selector type? Not sure, but it would need to be something that maybe has a counter on each card clicked, to track how many times you've selected it.

For now, this solution seems the most elegant & responsive to me.

File: server/game/cards/22-BtB/Highgarden.js (hunk 17 above)
Comment (Patane97) [reply server/game/cards/22-BtB/Highgarden.js]: Fair enough, however I decided to slightly alter this and use the same logic we have for many dynamic effects, where its stored in an object tying the effect to each card uuid affected. In this case, the property name is the STR increase (2 or 4), and the value is an array of characters its affected. This is part of the 63dd48b update; hopefully it suffices.
